Integrate religion data to demographic section

amitness opened the pull request.

amitness committed:
Integrate religion date to demographic section

cliftonmcintosh commented:
There appears to be a duplicate record. I received this error when populating the tables:

```
ERROR:  could not create unique index "religion_pkey"
DETAIL:  Key (geo_level, geo_code, religion_name)=(local, 46006, Prakriti) is duplicated.
```

I think local unit 46006 is duplicated, not just fro Prakriti.

cliftonmcintosh commented on sql/csv/population-by-religion.csv:
```diff
@@ -0,0 +1,9296 @@
+Hindu,401702,58006,local
+Islam,11084,58006,local
+Jainism,0,58006,local
+Kirat,42
```
new lines at the end of files, please

cliftonmcintosh commented on sql/religion.sql:
```diff
@@ -0,0 +1,61 @@
+
+--
+-- PostgreSQL database dump complete
+--
```
new line at the end of the file, please

cliftonmcintosh commented on wazimap_np/tables.py:
```diff
@@ -1,8 +1,17 @@
+           universe='Population',
+           description='Population by Religion',
+           dataset='National Population and Housing Census 2011',
            year='2011')
```
new line at the end of the file, please

cliftonmcintosh commented on wazimap_np/templates/profile/sections/demographics.html:
```diff
@@ -9,12 +9,29 @@ <h1>Demographics</h1>
         </div>
     {% endif %}
-</article>
+</article>
```
new line at the end of the file, please

amitness commented:
@cliftonmcintosh Thanks for the catch. There are around 57 such levels which got twice/thrice duplicates because of mapping based on the similarity of name.  I looked into the issue and looks like there are new local levels which have no equivalent levels in our geography.sql yet. 

I'm thinking of going an alternate route. Since we only have the population data right now in NepalMap federal and that was scraped from some other source, we could just consider the national portal's mapping from the province -> district -> local levels as the ground truth and regenerate new geography.sql with new ids based on its structure. The national portal also has 18 data which also includes the population by gender. This would be smoother for me to import without requiring any ETL and manual efforts in mapping.

For any future data loads after this, we can write a mapping script to map discrepancies to the national portal's hierarchical ids.

cliftonmcintosh commented:
If you re-generate the geography.sql, you will also need to re-generate the geo files -- the .topojson and .geojson files. The geography.sql and geo files must match. The data we present must be for the same geographical units as are present in the geo files.

amitness commented:
Oh yes. This regeneration idea is not going to work then, since I don't have topojson for national portal data. Looks like manually going through each duplicate and fixing mapping is the only way

amitness commented:
I was working on manually checking and fixing these duplicated ids, and came across a piece of interesting news. This was the cause for many of the ids. 
http://english.ratopati.com/story/2436
https://myrepublica.nagariknetwork.com/news/chief-wants-rural-municipality-named-after-his-dead-wife/
https://myrepublica.nagariknetwork.com/news/parties-renaming-public-places-after-their-own/

amitness committed:
Fix duplicated ids when mapping between local levels of NepalMap and National Portal

amitness commented:
@cliftonmcintosh I've fixed the issue. Can you re-check and merge it?

cliftonmcintosh commented:
@amitness 

Thank you for the updates. The duplicates are gone. I have been doing other analysis of the data to make sure that it conforms to what the Wazimap framework requires. There is one thing I see that will need updating. Every geographic unit needs to have data. Right now there are 30 units that do not. Without this, clicks on those units on the map will blow up. Most of them are parks. For those, we can can have zeroes since a decision was made to treat them as officially uninhabited. For the districts that are missing data, we shoud be able to get them by aggregating the data for all of their local units.

Units that are missing religion data
```sql
wazimap_np=# select id, geo_level, geo_code, name, parent_code from wazimap_geography where geo_code not in (select r.geo_code from religion r, wazimap_geography wg where r.geo_code = wg.geo_code and r.geo_level = wg.geo_level) order by geo_level, parent_code, geo_code;
 id  | geo_level | geo_code |                   name                   | parent_code 
-----+-----------+----------+------------------------------------------+-------------
  16 | district  | 08       | Bara                                     | 2
  11 | district  | 03       | Baglung                                  | 4
  10 | district  | 02       | Arghakhanchi                             | 5
  15 | district  | 07       | Banke                                    | 5
  17 | district  | 09       | Bardiya                                  | 5
   9 | district  | 01       | Achham                                   | 7
  12 | district  | 04       | Baitadi                                  | 7
  13 | district  | 05       | Bajhang                                  | 7
  14 | district  | 06       | Bajura                                   | 7
 764 | local     | 69099    | Khaptad National Park                    | 01
 513 | local     | 45099    | Dhorpatan Hunting Reserve                | 03
 753 | local     | 68099    | Khaptad National Park                    | 05
 740 | local     | 67099    | Khaptad National Park                    | 06
 396 | local     | 33099    | Parsa Wildlife Reserve                   | 08
 654 | local     | 58099    | Bardiya National Park                    | 09
 419 | local     | 35099    | Chitawan National Park                   | 12
 781 | local     | 70099    | Khaptad National Park                    | 22
 804 | local     | 72099    | Shuklaphanta National Park               | 32
 360 | local     | 31088    | Parsa Wildlife Reserve                   | 41
 361 | local     | 31099    | Chitawan National Park                   | 41
 495 | local     | 43099    | Dhorpatan Hunting Reserve                | 46
 839 | local     | 76099    | Chitawan National Park                   | 47
 320 | local     | 28088    | Shivapuri Watershed And Wildlife Reserve | 48
 321 | local     | 28099    | Langtang National Park                   | 48
 411 | local     | 34099    | Chitwan National Park                    | 53
 843 | local     | 77099    | Dhorpatan Hunting Reserve                | 59
 559 | local     | 49099    | Lumbini Sanskritik Development Area      | 60
 157 | local     | 15099    | Koshi Tappu Wildlife Reserve             | 63
 687 | local     | 6099     | Koshi Tappu Wildlife Reserve             | 69
 138 | local     | 14099    | Koshi Tappu Wildlife Reserve             | 75
(30 rows)
```

amitness commented:
@cliftonmcintosh Thank you for clarifying that. For national parks, yes I agree we should just set them to zero and hide section at the code level.

Regarding the districts, all of them have data in the CSV. I'm not sure why you're getting some districts without data in your query. For example, 'Bara' is mentioned in your SQL output, put it's geo code "8" is already present in the output CSV file for religion.

![image](https://user-images.githubusercontent.com/8587189/61589366-8d10ad00-abc8-11e9-9fbf-f13e764e8781.png)

amitness commented:
@cliftonmcintosh I want to propose a different way to handle cases for those levels. Instead of setting for each demographic data to 0 for these regions, we could handle it at the code level itself using builtin exceptions provided by Wazimap.
![image](https://user-images.githubusercontent.com/8587189/61590206-e1ba2500-abd4-11e9-82de-5fc4f0d75b80.png)

If the population is zero for a level, you already have code that shows this and so none of the demographic graph stuff on the frontend is called for missing levels.
![image](https://user-images.githubusercontent.com/8587189/61590211-f5fe2200-abd4-11e9-85d3-9223cb4efb3d.png)

amitness committed:
Handle case when we have no data for the provided geography code

amitness committed:
Add data for agriculture land use area and handle condition for geography area which have no corresponding data

amitness commented:
@cliftonmcintosh I've handled the missing ids in code and also added a new chart for agriculture area. Can you please review it and then merge it if everything is OK?

amitness committed:
Add data for agriculture land use area and handle condition for geography area which have no corresponding data

cliftonmcintosh commented:
@amitness 

I have some comments and questions.

**Missing districts in religion (and agriculture) data**
Thank you for the updates. I believe I have found a reason for why those districts are not represented in the religion data. The districts not represented have `geo_code`s of "01" to "09". The religion CSV does not include the "0". It has only single digits, "1" through "9". Therefore, the ingestion of the CSV fails to map those data to the correct district. In fact, rows are written in the database for those districts, but the `geo_codes` of "1" through "9" are incorrect. When NepalMap tries to find the data to display for those districts, it won't find anything.

If you update the CSV to include a "0" as part of the `geo_code` for those districts, then the data will be properly placed in the database, and the data will be found correctly.

The agriculture data for districts "01" through "09" have the same problem.

**Handling missing data programmatically**
Your suggestion to handle missing data programmatically by catching errors eliminates the need to have to make sure we have data for every entity, but I think that introduces a danger. The danger is that we won't notice when we are inadvertently missing data. For example, that would mask the incorrect `geo_code`s for districts with `geo_code`s of "01" through "09". For those districts, people would see a message that there is no data when, in fact, we have an error in our data ingestion.

Every time we create a new data set, there will be a possibility that we miss a geographic unit, and we won't notice. I think the errors will multiply. Personally, I think it is better to force ourselves to remember always add zeroes for the set of geographic units that we know have no data. It removes the chance of inadvertent errors creeping in without us noticing them. If we choose to suppress the errors, I think we will regret this later. It will increase the possibility that our data are incorrect.

These next items are not as crucial:

**Focused pull requests**
Would you be willing to separate the agriculture data into a separate pull request? It is generally better to have a pull request focus on one thing. 

1. It makes it easier to analyze.
2. When one part needs correcting (like the district `geo_code`s for religion), it doesn't then block the other thing that might be okay.
3. When people look back to try to find changes, it makes it easier to find them. For example, if someone is looking back through PRs to try to find where the agriculture data is added, they wouldn't expect to search in the religion data PR to find it.

**Making data transformation visible**
Is there a project or repository where the tools you used to create the CSVs are kept. One advantage of making that visible is that it allows people to see how data were transformed and makes the transformation process transparent. If someone asks, "How did you get those numbers?", we can say, "Here, this is the code that was used to get those numbers."

**Column naming**
This is very small and not crucial, but we should consider trying to be consistent about how we name the data columns. In the `religion` table, the data is `religion_name` (a code-y sort of name), while the data in agriculture is "Agricultural land use area" (a more natural sort of name). In the pre-federal version, we have used more natural names, with all lowercase words, like "percent deliveries at institutions" or "number of registered political parties". I have a slight preference for more natural names, but either way is fine with me. I would just suggest we be consistent.

cliftonmcintosh commented on wazimap_np/demographics.py:
```diff
@@ -20,4 +16,14 @@ def get_demographics_profile(geo_code, geo_level, session):
+        most_populous_religion = religion_data[religion_data.keys()[0]]
+        demographic_data['religion_distribution'] = religion_data
+        demographic_data['most_populous_religion'] = most_populous_religion
+    except LocationNotFound:
```
As I mentioned in my comment (https://github.com/Code4Nepal/nepalmap_federal/pull/62#issuecomment-513556766), this has the danger of masking inadvertent errors. When our data ingestion fails, this will make it less obvious that we have a problem. Requiring us to include rows for the parks and other areas where there is no official population is more purposeful and won't mask errors in other data.

cliftonmcintosh commented on wazimap_np/tables.py:
```diff
@@ -1,8 +1,24 @@
+           universe='Agriculture',
+           description='Agriculture Land Use Area',
+           dataset='National Portal',
            year='2011')
```
new lines at ends of files is preferred.

cliftonmcintosh commented on wazimap_np/templates/profile/sections/agriculture.html:
```diff
@@ -0,0 +1,25 @@
+            </section>
+        </div>
+    {% endif %}
+</article>
```
new lines at ends of files is preferred.

amitness replied:
Sure. The IDE I'm using does this for Python files automatically but doesn't do it for HTML.

cliftonmcintosh commented:
@amitness 

Is there a way to have a whitelist of geographic units that we know don't have data and handle those differently? In order of preference, here is what I think the options are:

1. Find a programmatic way to deliver a "no data" message for specific geographic units that we know have no official data.
2. Insert zeroes for all data points for the specifici geographic units that we know have no official data.
3. Always catch errors when there is no data for a specific area (this is the choice currently implemented in this PR).

As mentioned, I think option 3 will mask errors, and I would prefer to see option 2 before we do that.

amitness commented:
@cliftonmcintosh One way could be to have a whitelist of (geo_code, geo_level) tuple and then catch-all location not found exceptions. Then, inside the except clause, check if current geo_code, geo_level is in whitelist or not. If it is, we pass the exception, else, we rethrow the exception. What do you think?

![image](https://user-images.githubusercontent.com/8587189/61592439-ef7ea300-abf2-11e9-93fd-bd7e34d07ff4.png)

cliftonmcintosh commented:
We should be able to do the check before the exception. We have an approach to that in the pre-federal version. See, for example, https://github.com/Code4Nepal/nepalmap_app/blob/dev/wazimap_np/agriculture.py#L45-L121.

Here is a sample of what the `demographics.py` file might look like if we adopted something similar. It would probably be good to move the set of geo_codes into a shared place so they can be referenced by all profile files and not be copied each time.

```python
from wazimap.data.utils import get_stat_data

park_geo_codes = {'6099', '14099', '15099', '28088', '28099', '31088', '31099', '33099', '34099', '35099', '43099', '45099', '49099', '58099', '67099', '68099', '69099', '70099', '72099', '76099', '77099'}


def get_demographics_profile(geo_code, geo_level, session):
    demographic_data = {'area_has_data': False}

    if should_have_data(geo_code, geo_level):
        pop_data, total_pop = get_stat_data(
            'sex', geo_level, geo_code, session,
            table_fields=['sex'],
            table_name='population')

        demographic_data = {
            'area_has_data': True,
            'pop_dist': pop_data,
            'has_data': total_pop > 0,
            'total_population': {
                'name': 'People',
                'values': {'this': total_pop}
            }
        }

        religion_data, _ = get_stat_data(['religion_name'], geo_level,
                                         geo_code, session,
                                         order_by='-total')
        most_populous_religion = religion_data[religion_data.keys()[0]]
        demographic_data['religion_distribution'] = religion_data
        demographic_data['most_populous_religion'] = most_populous_religion

    return demographic_data


def should_have_data(geo_code, geo_level):
    return geo_level != 'local' and geo_code not in park_geo_codes
```

cliftonmcintosh commented:
@amitness 

There are errors when using these data as it is currently done. For example, when I choose a local area in the Accham district, it fails to display: http://localhost:8000/profiles/local-69005-mangalsen/

Similar problem with a local unit in Baitadi. See, for example, http://localhost:8000/profiles/local-74009-sigas/.

I believe this is related to the district data being unavailable because of the problem with no zeroes at the beginning of the geo_code for each of those districts. When the information about the locality is retrieved, it also tries to get information about the levels above it for reporting, and it fails because it can't find the religion data for the district.

FWIW, the lack of religion data for parks seems to be okay. Here is an example of a table that gets displayed for Doti district, which has Khaptad National Park in it: http://localhost:8000/data/table/?table=RELIGION&geo_ids=local-70002,district-22,province-7,country-NP,local|district-22&primary_geo_id=local-70002

Noticed that the park has "N/A" for its data, but data for the rest of the localities are displayed.

Screenshot of error for district with no religion data:
![show-data-error](https://user-images.githubusercontent.com/3824492/61593485-c5eb5a80-aba5-11e9-9604-323f8182b75b.png)

amitness committed:
Fix district code issue and regenerate CSV files for religion and agriculture data

amitness committed:
Handle missing data for selected list of national parks and local levels

amitness commented:
@cliftonmcintosh I've updated the code to fix the issue you mentioned above. Please review them.

**Fixes**
1. District code from zero to nine should be two digits: '01', '02' ... '09'
Should work: 
http://localhost:8000/profiles/local-69005-mangalsen/
http://localhost:8000/profiles/local-74009-sigas/

2. Removed exceptions to handle missing data and instead use a blacklist of geocodes for which, data missing message should be displayed. 

3. Updates column names to be lowercase and natural sounding 
![image](https://user-images.githubusercontent.com/8587189/62845624-50c4fe00-bce9-11e9-8fc0-c7dfe71df700.png)

amitness committed:
Use lowercase natural names for columns in tables

cliftonmcintosh commented:
@amitness 

All localities are now displaying "No data available for this area" for me. Perhaps there is something wrong with the logic for displaying "No data available for this area".

Here is a screenshot of Lalitpur:
![lalitpur-no-data](https://user-images.githubusercontent.com/3824492/62880222-89b5a080-bcf2-11e9-91f0-4a937337d294.png)


And here is a query showing Lalitpur has population data:
```sql
 geo_level | geo_code |  sex   | total  
-----------+----------+--------+--------
 local     | 25004    | male   | 145924
 local     | 25004    | female | 138998
(2 rows)
```

cliftonmcintosh commented on wazimap_np/profiles.py:
```diff
@@ -1,60 +1,61 @@
+
+
+def should_have_data(geo_code, geo_level):
+    return geo_level != 'local' and geo_code not in park_geo_codes
```
The sample `should_have_data` method I provided in my earlier comment was incorrect. It evaluates to `False` for all local areas because `geo_level != 'local'` always evaluates to `False` for local areas. Does or work instead of and? As in:

```python
def should_have_data(geo_code, geo_level):
   return geo_level != 'local' or geo_code not in park_geo_codes
```

I believe so. After I change the `and` to `or`, I see data:

![lalitpur-data](https://user-images.githubusercontent.com/3824492/62881888-6987e080-bcf6-11e9-8e53-86a912bdc203.png)

amitness replied:
I thought of something different but looks like what you proposed and what I've done is essentially the same thing. 

My approach:
```python
return not (geo_level == 'local' and geo_code in park_geo_codes)
```

By DeMorgan's law, when you apply `not` to an expression, `and` becomes `or`.
So,
```python
return geo_level != 'local' or geo_code not in park_geo_codes
```
This is what you proposed.

amitness committed:
Fix issues in displaying data of local levels due to blacklist

amitness commented:
@cliftonmcintosh I've pushed a fix. Please check now.

cliftonmcintosh commented:
Spot-checking looks good.

amitness commented:
@cliftonmcintosh Thanks for reviewing. Does the changes autodeploy to live federal site or is it a manual process?

cliftonmcintosh commented:
It does not auto-deploy. I am in the process of updating the server for beta.nepalmap.org right now. I encountered a difficulty that I haven't encountered before, and I'm troubleshooting.

cliftonmcintosh commented:
@amitness 

Your updates have been deployed and should be visible at https://beta.nepalmap.org
